File: android/src/main/kotlin/com/imf/blade/BladePlugin.kt
package com.imf.blade

import androidx.annotation.NonNull
import com.imf.blade.container.events.BackgroundEvent
import com.imf.blade.container.events.ForegroundEvent
import com.imf.blade.container.nativeEvents.*
import com.imf.blade.messager.FlutterChannel

import io.flutter.embedding.engine.plugins.FlutterPlugin

/** BladePlugin */
class BladePlugin: FlutterPlugin, NativeEventListener {
  var flutterContainerManager = FlutterContainerManager()
  lateinit var flutterChannel: FlutterChannel
  lateinit var platformCoordinator: PlatformCoordinator

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    val binaryMessenger = flutterPluginBinding.binaryMessenger
    val name = "com.imf.blade"
    flutterChannel = FlutterChannel(binaryMessenger, name)
    platformCoordinator = PlatformCoordinator(binaryMessenger, flutterContainerManager, this)
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    platformCoordinator.release()
  }

  var delegate: NativeEventListener? = null

  override fun pushFlutterPage(event: PushFlutterPageEvent) {
    delegate?.pushFlutterPage(event)
  }

  override fun pushNativePage(event: PushNativePageEvent) {
    delegate?.pushNativePage(event)
  }

  override fun popNativePage(event: PopNativePageEvent) {
    delegate?.popNativePage(event)
  }

  override fun popUntilNativePage(event: PopUntilNativePageEvent) {
    delegate?.popUntilNativePage(event)
  }

  fun handleForegroundEvent() {
    flutterChannel.sendEvent(ForegroundEvent())
  }

  fun handleBackgroundEvent() {
    flutterChannel.sendEvent(BackgroundEvent())
  }
}

File: ios/Classes/BladePlugin.swift
import Flutter
import UIKit

public class BladePlugin: NSObject, FlutterPlugin,NativeEventListener {
    
    var delegate: BladeDelegate?
    var channel: FlutterChannel?
    var coordinator:PlatformCoordinator?
    var containerManager:FlutterContainerManager?
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.imf.blade", binaryMessenger: registrar.messenger())
        let instance = BladePlugin()
        instance.containerManager = FlutterContainerManager()
        instance.coordinator = PlatformCoordinator(channel: channel)
        instance.channel = FlutterChannel(messager: registrar.messenger(),channel: channel)
        registrar.publish(instance)
    }

    public func pushFlutterPage(event: FlutterBaseEvent) {
        delegate?.pushFlutterPage(event: event)
    }
    
    public func pushNativePage(event: FlutterBaseEvent) {
        delegate?.pushNativePage(event: event)
    }
    
    static func getPlugin(engine:FlutterEngine?) -> BladePlugin?{
        let published = engine?.valuePublished(byPlugin: "BladePlugin")
        if let plugin = published as? BladePlugin {
            return plugin
        }
        return nil
    }
    
    static var plugin: BladePlugin? {
        return BladePlugin.getPlugin(engine: Blade.shared.engine)
    }
}
